presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java

```diff
@@ -0,0 +1,507 @@
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.facebook.presto.sql.tree.TryExpression;
+import com.facebook.presto.sql.tree.WhenClause;
+import io.airlift.slice.Slice;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.spi.function.OperatorType.ADD;
+import static com.facebook.presto.spi.function.OperatorType.DIVIDE;
+import static com.facebook.presto.spi.function.OperatorType.EQUAL;
+import static com.facebook.presto.spi.function.OperatorType.GREATER_THAN;
+import static com.facebook.presto.spi.function.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.spi.function.OperatorType.IS_DISTINCT_FROM;
+import static com.facebook.presto.spi.function.OperatorType.LESS_THAN;
+import static com.facebook.presto.spi.function.OperatorType.LESS_THAN_OR_EQUAL;
+import static com.facebook.presto.spi.function.OperatorType.MODULUS;
+import static com.facebook.presto.spi.function.OperatorType.MULTIPLY;
+import static com.facebook.presto.spi.function.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.spi.function.OperatorType.SUBTRACT;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.COALESCE;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IN;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.SWITCH;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.WHEN;
+import static com.facebook.presto.sql.planner.RowExpressionInterpreter.rowExpressionInterpreter;
+import static com.facebook.presto.sql.relational.StandardFunctionResolution.getArithmeticOperator;
+import static com.facebook.presto.sql.relational.StandardFunctionResolution.getComparisonOperator;
+import static com.facebook.presto.sql.relational.StandardFunctionResolution.isArithmeticFunction;
+import static com.facebook.presto.sql.relational.StandardFunctionResolution.isComparisonFunction;
+import static com.facebook.presto.sql.tree.LogicalBinaryExpression.Operator.AND;
+import static com.facebook.presto.sql.tree.LogicalBinaryExpression.Operator.OR;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * RowExpression visitor which verifies if given expression (actual) is matching other RowExpression given as context (expected).
+ */
+final class RowExpressionVerifier
+        extends AstVisitor<Boolean, RowExpression>
+{
+    // either use variable or input reference for symbol mapping
+    private final SymbolAliases symbolAliases;
+    private final Metadata metadata;
+    private final Session session;
+    private final StandardFunctionResolution functionResolution;
+
+    RowExpressionVerifier(SymbolAliases symbolAliases, Metadata metadata, Session session)
```

Comment:
Can we convert Expression to RowExpression first and verify based on RowExpression?  This is used in test only, so probably we want to catch verifier.

Reply:
I thought about that; but it requires support type provider for all the users of `PlanMatchPattern`. There are way too many of them. Given it is only for plan matching, I would keep it simple without types.